bokehjs/src/lib/models/glyphs/glyph.ts

```diff
@@ -238,97 +234,59 @@ export abstract class GlyphView extends View {
       }
-      data = data_subset
-    }
 
-    let data_size = Infinity
-    for (const k in data) {
-      if (k.charAt(0) === '_')
-        data_size = Math.min(data_size, (data[k] as any).length)
+      Object.defineProperty(this, `_${name}`, {
+        value: array,
+        configurable: true,
+      })
     }
-    if (data_size != Infinity)
-      this._data_size = data_size
-    else
-      this._data_size = 0 // XXX: this only happens in degenerate unit tests
-
-    const self = this as any
-    extend(self, data)
 
-    // TODO (bev) Should really probably delegate computing projected
-    // coordinates to glyphs, instead of centralizing here in one place.
     if (this.renderer.plot_view.model.use_map) {
-      if (self._x != null)
-        [self._x, self._y] = proj.project_xy(self._x, self._y)
-
-      if (self._xs != null)
-        [self._xs, self._ys] = proj.project_xsys(self._xs, self._ys)
-
-      if (self._x0 != null)
-        [self._x0, self._y0] = proj.project_xy(self._x0, self._y0)
-
-      if (self._x1 != null)
-        [self._x1, self._y1] = proj.project_xy(self._x1, self._y1)
-    }
-
-    function num_array(array: Arrayable<number>): NumberArray {
-      if (array instanceof NumberArray)
-        return array
-      else
-        return new NumberArray(array)
-    }
-
-    // if we have any coordinates that are categorical, convert them to
-    // synthetic coords here
-    const xr = this.renderer.scope.x_range
-    const yr = this.renderer.scope.y_range
-
-    // XXX: MultiPolygons is a special case of special cases
-    if (this.model.type != "MultiPolygons") {
-      for (let [xname, yname] of this.model._coords) {
-        xname = `_${xname}`
-        yname = `_${yname}`
-
-        // TODO (bev) more robust detection of multi-glyph case
-        // hand multi glyph case
-        if (self._xs != null) {
-          if (xr instanceof FactorRange)
-            self[xname] = map(self[xname], (arr: any) => xr.v_synthetic(arr))
-          else
-            self[xname] = map(self[xname], num_array)
-          if (yr instanceof FactorRange)
-            self[yname] = map(self[yname], (arr: any) => yr.v_synthetic(arr))
-          else
-            self[yname] = map(self[yname], num_array)
-        } else {
-          // hand standard glyph case
-          if (xr instanceof FactorRange)
-            self[xname] = xr.v_synthetic(self[xname])
-          else
-            self[xname] = num_array(self[xname])
-          if (yr instanceof FactorRange)
-            self[yname] = yr.v_synthetic(self[yname])
-          else
-            self[yname] = num_array(self[yname])
-        }
-      }
+      this._project_data()
     }
 
-    if (this.glglyph != null)
-      this.glglyph.set_data_changed(self._x.length)
+    this.glglyph?.set_data_changed()
 
-    this._set_data(indices_to_update)  //TODO doesn't take subset indices into account
+    this._set_data(indices_to_update)  // TODO doesn't take subset indices into account
```

Comment:
FWIW I have no idea what this comment means anymore (if I ever did)

Reply:
I'm not sure about this one exactly, but indexing is a bit of a mess right now, and a performance hog (e.g. constant conversions between subset and full space indices often take 10x the time of searching the spatial index). This may get addressed in PR #10229, or on top of it.